code/controllers/subsystems/zcopy.dm

```diff
@@ -114,82 +109,91 @@
 				break
 			continue
 
-		if (!T.shadower)	// If we don't have our shadower yet, create it.
-			T.shadower = new(T)
+		if (!T.shadower)
+			WARNING("Turf [T] at [T.x],[T.y],[T.z] was queued, but had no shadower.")
+			continue
 
 		// Figure out how many z-levels down we are.
 		var/depth = 0
 		var/turf/Td = T
-		while (Td && TURF_IS_MIMICING(Td.below))
-			Td = Td.below
-			depth++
+
+		while (Td.above)
+			Td = Td.above
+			depth += 1
+
 		if (depth > OPENTURF_MAX_DEPTH)
 			depth = OPENTURF_MAX_DEPTH
 
-		var/oo_target = OPENTURF_MAX_PLANE - depth
-		var/t_target
+		var/t_target = OPENTURF_MAX_PLANE	// this is where the openturf gets put
 
-		// Handle space parallax & starlight.
-		if (T.is_above_space())
-			t_target = PLANE_SPACE_BACKGROUND
-			if (starlight_enabled && !T.light_range)
-				T.set_light(config.starlight, 0.5)
-		else
-			t_target = oo_target
-			if (starlight_enabled && T.light_range)
-				T.set_light(0)
+		// Handle space parallax.
+		if (T.below.z_eventually_space)
+			T.z_eventually_space = TRUE
 
-		if (!(T.flags & MIMIC_OVERWRITE))
+			if (istype(T.below, /turf/space))
+				t_target = PLANE_SPACE
+
+		if (!(T.flags & ZM_MIMIC_OVERWRITE))
 			// Some openturfs have icons, so we can't overwrite their appearance.
 			if (!T.below.bound_overlay)
 				T.below.bound_overlay = new(T)
 			var/atom/movable/openspace/turf_overlay/TO = T.below.bound_overlay
 			TO.appearance = T.below
 			TO.name = T.name
 			TO.opacity = FALSE
-			T.desc = TO.desc = "Below seems to be \a [T.below]."
 			TO.plane = t_target
-			TO.mouse_opacity = FALSE
 		else
 			// This openturf doesn't care about its icon, so we can just overwrite it.
 			if (T.below.bound_overlay)
 				QDEL_NULL(T.below.bound_overlay)
 			T.appearance = T.below
 			T.name = initial(T.name)
+			T.desc = initial(T.desc)
 			T.gender = NEUTER
 			T.opacity = FALSE
-			T.plane = t_target
+			T.plane = t_target - depth
 
-		T.desc = "Below seems to be \a [T.below]."
 		T.queue_ao()	// No need to recalculate ajacencies, shouldn't have changed.
 
 		// Add everything below us to the update queue.
 		for (var/thing in T.below)
 			var/atom/movable/object = thing
-			if (QDELETED(object) || object.no_z_overlay || object.loc != T.below)
-				// Don't queue deleted stuff or stuff that doesn't need an overlay.
+			if (QDELETED(object) || object.no_z_overlay || object.loc != T.below || object.invisibility == INVISIBILITY_ABSTRACT)
+				// Don't queue deleted stuff, stuff that's not visible, blacklisted stuff, or stuff that's centered on another tile but intersects ours.
 				continue
 
-			if (object.type == /atom/movable/lighting_overlay)	// Special case.
-				T.shadower.copy_lighting(object)
+			// Special case: these are merged into the shadower to reduce memory usage.
+			if (object.type == /atom/movable/lighting_overlay)
+				// T.shadower.copy_lighting(object)
```

Comment:
The existing aurora copy_lighting implementation works fine, no need to comment it out.